### PR TITLE
[SYSTEMDS-2550] Federated Parameter Server

### PR DESCRIPTION
**Pull request by tobiasrieger**
If a federated input Matrix is passed to the parameter server builtin instruction, the parameter server will detect it. 

**Data Partitioning**
for now only row federated matrices are supported. The federated data partitioner extracts the information about the workers.
**Setup**
The setup is handled by a "User Defined Function" (UDF). To send all necessary data (model, gradient and aggregation functions) to the federated worker it was required to serialise ListObjects and parts of the program. As the parameter server uses unoptimised functions, it was further necessary to extend the program serialisation functionality of the ProgramConverter.
**Training**
There are two different update types and frequencies for the federated workers. Updates are possible "batch synchronised" or "asynchronous". An update can be sent each batch or epoch. Each computation of an update is handled by a UDF which is sent along with the current model. In the epoch case, the federated worker can also use the aggregation function to update its local model.
**Cleanup**
Most cleanup is done on the fly and making sure to clear the caches for "Cacheable Objects". After the federated worker completes its task, a "teardown UDF" is sent to restore the execution context to the state before the computations.
**Testing**
The included test case, tests all different supported parameter combinations on a randomly generated "dummy MNIST" dataset.

**Problems**
At the moment the tests only work for the Exec Mode "Single Node". To fix this the program is recompiled on call of the parameter server builtin instruction, but this process creates invalid CP°rmvar instructions when passing SPARK instructions

### CODE REVIEW

**Review by sebwrede — COMMENTED** [review]
The PR looks good to me. 
I added some comments regarding privacy constraints. It is up to you, if you want to do it in this PR since it is not essential to the feature you are implementing. I will anyway look at it later.

File: src/test/java/org/apache/sysds/test/component/paramserv/SerializationTest.java [hunk 1]
```diff
@@ -29,40 +35,90 @@
...
+		try {
+			ByteArrayOutputStream bos = new ByteArrayOutputStream();
+			ObjectOutputStream out = new ObjectOutputStream(bos);
+			out.writeObject(lo);
+			out.flush();
+			byte[] loBytes = bos.toByteArray();
+
+			ByteArrayInputStream bis = new ByteArrayInputStream(loBytes);
+			ObjectInput in = new ObjectInputStream(bis);
+			loDeserialized = (ListObject) in.readObject();
```
[the remark this answers is not in the document]
Comment (tobiasrieger) [reply src/test/java/org/apache/sysds/test/component/paramserv/SerializationTest.java:82]: Thanks 😁

File: src/main/java/org/apache/sysds/runtime/controlprogram/paramserv/dp/KeepDataOnWorkerFederatedScheme.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	}
+
+	/**
+	 * Takes a row federated Matrix and slices it into a matrix for each worker
+	 *
+	 * @param fedMatrix the federated input matrix
+	 */
+	private List<MatrixObject> sliceFederatedMatrix(MatrixObject fedMatrix) {
+		if (fedMatrix.isFederated(FederationMap.FType.ROW)
+				|| fedMatrix.isFederated(FederationMap.FType.ROW)) {
```
[the remark this answers is not in the document]
Comment (tobiasrieger) [reply src/main/java/org/apache/sysds/runtime/controlprogram/paramserv/dp/KeepDataOnWorkerFederatedScheme.java:53]: By mistake

File: src/main/java/org/apache/sysds/runtime/instructions/cp/ParamservBuiltinCPInstruction.java [hunk 3]
```diff
@@ -91,16 +95,87 @@ public ParamservBuiltinCPInstruction(Operator op, LinkedHashMap<String, String>
...
+				.namingPattern("workers-pool-thread-%d").build();
+		ExecutorService es = Executors.newFixedThreadPool(workerNum, factory);
+
+		// Get the compiled execution context
+		LocalVariableMap newVarsMap = createVarsMap(ec);
+		// Level of par is 1 because one worker will be launched per task
+		// TODO: Fix recompilation
+		ExecutionContext newEC = ParamservUtils.createExecutionContext(ec, newVarsMap, updFunc, aggFunc, 1, true);
+		// Create workers' execution context
+		List<ExecutionContext> federatedWorkerECs = ParamservUtils.copyExecutionContext(newEC, workerNum);
```
[the remark this answers is not in the document]
Comment (tobiasrieger) [reply src/main/java/org/apache/sysds/runtime/instructions/cp/ParamservBuiltinCPInstruction.java:147]: That is a little more complicated. Each FederatedPSControlThread needs its own execution context for synchronisation. The FederatedPSControlThread then sets up everything the federated worker needs in its execution context. 

To go back to the question: The only redundant thing is, that the EC of the FedPSThread does not need code other than the gradients and aggregation function, but for now there is no filtering. It gets filtered when serialising and sending to the federated worker though.

File: src/main/java/org/apache/sysds/runtime/instructions/cp/ListObject.java [hunk 4]
```diff
@@ -280,4 +301,124 @@ public String toString() {
...
+					out.writeObject(dc.getRows());
+					out.writeObject(dc.getCols());
+					out.writeObject(dc.getBlocksize());
+					out.writeObject(dc.getNonZeros());
+					out.writeObject(md.getFileFormat());
+					out.writeObject(mo.acquireReadAndRelease());
+					break;
+				case SCALAR:
+					ScalarObject so = (ScalarObject) d;
+					out.writeObject(so.getStringValue());
```
[the remark this answers is not in the document]
Comment (tobiasrieger) [reply src/main/java/org/apache/sysds/runtime/instructions/cp/ListObject.java:351]: Thanks for the input! Added 👍🏻

File: src/main/java/org/apache/sysds/runtime/instructions/cp/ListObject.java [hunk 5]
```diff
@@ -280,4 +301,124 @@ public String toString() {
...
+					int blockSize = (int) in.readObject();
+					long nonZeros = (long) in.readObject();
+					Types.FileFormat fileFormat = (Types.FileFormat) in.readObject();
+
+					// construct objects and set meta data
+					MatrixCharacteristics matrixCharacteristics = new MatrixCharacteristics(rows, cols, blockSize, nonZeros);
+					MetaDataFormat metaDataFormat = new MetaDataFormat(matrixCharacteristics, fileFormat);
+					MatrixBlock matrixBlock = (MatrixBlock) in.readObject();
+
+					d = new MatrixObject(valueType, Dag.getNextUniqueVarname(Types.DataType.MATRIX), metaDataFormat, matrixBlock);
```
[the remark this answers is not in the document]
Comment (tobiasrieger) [reply src/main/java/org/apache/sysds/runtime/instructions/cp/ListObject.java:403]: I have realised, that all data objects have PCs, so I have just generalised it. Take a look at the new commit 😊

**Comment by Baunsgaard** [comment]
does the removal of WIP, mean it is ready for review?

**Comment by Baunsgaard** [comment]
could you commit something again to make the tests run, i have fixed master so that it should not break like in the last commit.

**Pull request closed without merging**